lms/djangoapps/grades/models.py

```diff
@@ -244,6 +244,35 @@ class Meta(object):
     # track which blocks were visible at the time of grade calculation
     visible_blocks = models.ForeignKey(VisibleBlocks, db_column='visible_blocks_hash', to_field='hashed')
 
+    def _is_unattempted_with_score(self):
+        """
+        Return True if the object has a non-zero score, but has not been
+        attempted.  This is an inconsistent state, and needs to be cleaned up.
+        """
+        return self.first_attempted is None and any(field != 0.0 for field in (self.earned_all, self.earned_graded))
+
+    def enforce_unattempted(self, save=True):
+        """
+        If an grade has not been attempted, but was given a non-zero score,
+        reset the score to 0.0.
+
+        Params:
+
+        save (bool, default: True):
+            By default, this method saves the model if and only if there was an
+            inconsistency.  If the caller needs to save the model regardless of
+            the result, or will be saving the model later after making other
+            changes, this may be an unwanted database request.  It can be
+            disabled by passing ``save=False``.
+
+        Return value: None
+        """
+        if self._is_unattempted_with_score():
```

Comment:
Why are we silently updating the model if the code is violating the expected invariant?  Instead, we should raise an error and fix the code.